File: dungeons-1.21.11/src/main/java/fishmod/features/CommandKeysScreen.java
package fishmod.features;

import fishmod.features.other.CommandKeys;
import net.minecraft.client.gui.Click;
import net.minecraft.client.gui.DrawContext;
import net.minecraft.client.gui.screen.Screen;
import net.minecraft.client.gui.widget.ButtonWidget;
import net.minecraft.client.gui.widget.TextFieldWidget;
import net.minecraft.client.input.KeyInput;
import net.minecraft.client.util.InputUtil;
import net.minecraft.text.Text;
import org.lwjgl.glfw.GLFW;

import java.util.ArrayList;
import java.util.List;

/**
 * /fm commandkeys — bind arbitrary keys/mouse buttons to slash commands.
 *
 * <p>Entries are kept as plain {@code keys}/{@code commands} lists (the source of truth) and
 * widgets are rebuilt from them on every add/remove/scroll/rebind — simplest way to support a
 * variable-length, scrollable row list without hand-rolled widget recycling.
 *
 * <p>Key capture mirrors {@link FishModScreen}'s existing keybind-rebind convention: click a key
 * box to arm capture, then the next key or mouse click is bound; Escape unbinds instead.
 */
public class CommandKeysScreen extends Screen {

    private static final int BG_PANEL   = 0xF20E1016;
    private static final int BG_SECTION = 0xFF171A22;
    private static final int BORDER     = 0xFF2A2D38;
    private static final int ACCENT     = 0xFF55FFFF;
    private static final int TEXT_HINT  = 0xFF8A8F9C;
    private static final int LIST_BG    = 0xFF14161D;

    private static final int ROW_H = 24;
    private static final int MAX_VISIBLE = 6;
    private static final int KEY_BTN_W = 120;
    private static final int REMOVE_BTN_W = 20;

    private final List<InputUtil.Key> keys = new ArrayList<>();
    private final List<String> commands = new ArrayList<>();
    private Integer capturingIndex = null;
    private int scroll = 0;

    private int panelX, panelY;
    private final int panelW = 400;
    private int panelH;
    private int listX, listY, listW, listH;
    private int cmdFieldX, cmdFieldW, removeBtnX;

    public CommandKeysScreen() { super(Text.literal("Command Keys")); }

    @Override
    protected void init() {
        for (CommandKeys.Entry e : CommandKeys.all()) {
            keys.add(e.key());
            commands.add(e.command());
        }

        listH = ROW_H * MAX_VISIBLE;
        panelH = 50 + listH + 46;
        panelX = (this.width - panelW) / 2;
        panelY = Math.max(8, (this.height - panelH) / 2);

        listX = panelX + 14;
        listY = panelY + 44;
        listW = panelW - 28;
        cmdFieldX = listX + KEY_BTN_W + 6;
        cmdFieldW = listW - KEY_BTN_W - 6 - REMOVE_BTN_W - 6;
        removeBtnX = listX + listW - REMOVE_BTN_W;

        rebuildRows();
    }

    private void persist() {
        List<CommandKeys.Entry> list = new ArrayList<>();
        for (int i = 0; i < keys.size(); i++) list.add(new CommandKeys.Entry(keys.get(i), commands.get(i)));
        CommandKeys.replaceAll(list);
    }

    private void rebuildRows() {
        clearChildren();

        for (int i = 0; i < keys.size(); i++) {
            int rowTop = listY + i * ROW_H - scroll;
            if (rowTop + ROW_H < listY || rowTop > listY + listH) continue;
            final int idx = i;

            InputUtil.Key k = keys.get(i);
            String label = (capturingIndex != null && capturingIndex == i)
                    ? "> Press a key <"
                    : (k.equals(InputUtil.UNKNOWN_KEY) ? "Unbound" : k.getLocalizedText().getString());
            addDrawableChild(ButtonWidget.builder(Text.literal(label), b -> {
                        capturingIndex = idx;
                        rebuildRows();
                    })
                    .dimensions(listX, rowTop + 3, KEY_BTN_W, 18).build());

            TextFieldWidget cmdField = new TextFieldWidget(this.textRenderer, cmdFieldX, rowTop + 3, cmdFieldW, 18, Text.literal("Command"));
            cmdField.setMaxLength(256);
            cmdField.setText(commands.get(i));
            cmdField.setChangedListener(s -> { commands.set(idx, s); persist(); });
            addDrawableChild(cmdField);

            addDrawableChild(ButtonWidget.builder(Text.literal("§cX"), b -> {
                        keys.remove(idx); commands.remove(idx);
                        persist(); rebuildRows();
                    })
                    .dimensions(removeBtnX, rowTop + 3, REMOVE_BTN_W, 18).build());
        }

        int btnY = listY + listH + 8;
        addDrawableChild(ButtonWidget.builder(Text.literal("+ Add Command Key"), b -> {
                    keys.add(InputUtil.UNKNOWN_KEY); commands.add("");
                    persist(); rebuildRows();
                })
                .dimensions(panelX + 14, btnY, 160, 20).build());
        addDrawableChild(ButtonWidget.builder(Text.literal("Done"), b -> close())
                .dimensions(panelX + panelW - 14 - 70, btnY, 70, 20).build());
    }

    @Override
    public void renderBackground(DrawContext ctx, int mouseX, int mouseY, float delta) {
        super.renderBackground(ctx, mouseX, mouseY, delta);
        ctx.fill(panelX - 1, panelY - 1, panelX + panelW + 1, panelY + panelH + 1, BORDER);
        ctx.fill(panelX, panelY, panelX + panelW, panelY + panelH, BG_PANEL);
        ctx.fill(panelX, panelY, panelX + panelW, panelY + 22, BG_SECTION);
        ctx.fill(panelX, panelY + 22, panelX + panelW, panelY + 23, ACCENT);
        ctx.drawCenteredTextWithShadow(this.textRenderer, "§b§lCommand Keys", panelX + panelW / 2, panelY + 7, 0xFFFFFF);
        ctx.drawTextWithShadow(this.textRenderer,
                "§7Click a key box, then press a key or click a mouse button §8(Esc to unbind)",
                panelX + 14, panelY + 30, TEXT_HINT);
        ctx.fill(listX - 2, listY - 2, listX + listW + 2, listY + listH + 2, LIST_BG);
    }

    @Override
    public boolean mouseClicked(Click click, boolean bl) {
        if (capturingIndex != null) {
            keys.set(capturingIndex, InputUtil.Type.MOUSE.createFromCode(click.button()));
            capturingIndex = null;
            persist();
            rebuildRows();
            return true;
        }
        return super.mouseClicked(click, bl);
    }

    @Override
    public boolean keyPressed(KeyInput input) {
        if (capturingIndex != null) {
            keys.set(capturingIndex, input.key() == GLFW.GLFW_KEY_ESCAPE
                    ? InputUtil.UNKNOWN_KEY
                    : InputUtil.fromKeyCode(input));
            capturingIndex = null;
            persist();
            rebuildRows();
            return true;
        }
        return super.keyPressed(input);
    }

    @Override
    public boolean mouseScrolled(double mouseX, double mouseY, double horizontalAmount, double verticalAmount) {
        int max = Math.max(0, keys.size() * ROW_H - listH);
        scroll = Math.max(0, Math.min(max, scroll - (int) (verticalAmount * ROW_H)));
        rebuildRows();
        return true;
    }
}

File: dungeons-1.21.11/src/main/kotlin/fishmod/features/CommandKeysScreen.kt
package fishmod.features

import fishmod.features.other.CommandKeys
import net.minecraft.client.gui.Click
import net.minecraft.client.gui.DrawContext
import net.minecraft.client.gui.screen.Screen
import net.minecraft.client.gui.widget.ButtonWidget
import net.minecraft.client.gui.widget.TextFieldWidget
import net.minecraft.client.input.KeyInput
import net.minecraft.client.util.InputUtil
import net.minecraft.text.Text
import org.lwjgl.glfw.GLFW
import kotlin.math.max
import kotlin.math.min

/**
 * /fm commandkeys — bind arbitrary keys/mouse buttons to slash commands.
 *
 * Entries are kept as plain `keys`/`commands` lists (the source of truth) and
 * widgets are rebuilt from them on every add/remove/scroll/rebind — simplest way to support a
 * variable-length, scrollable row list without hand-rolled widget recycling.
 *
 * Key capture mirrors [FishModScreen]'s existing keybind-rebind convention: click a key
 * box to arm capture, then the next key or mouse click is bound; Escape unbinds instead.
 */
class CommandKeysScreen : Screen(Text.literal("Command Keys")) {

    companion object {
        private const val BG_PANEL = 0xF20E1016.toInt()
        private const val BG_SECTION = 0xFF171A22.toInt()
        private const val BORDER = 0xFF2A2D38.toInt()
        private const val ACCENT = 0xFF55FFFF.toInt()
        private const val TEXT_HINT = 0xFF8A8F9C.toInt()
        private const val LIST_BG = 0xFF14161D.toInt()

        private const val ROW_H = 24
        private const val MAX_VISIBLE = 6
        private const val KEY_BTN_W = 120
        private const val REMOVE_BTN_W = 20
    }

    private val keys: MutableList<InputUtil.Key> = ArrayList()
    private val commands: MutableList<String> = ArrayList()
    private var capturingIndex: Int? = null
    private var scroll = 0

    private var panelX = 0
    private var panelY = 0
    private val panelW = 400
    private var panelH = 0
    private var listX = 0
    private var listY = 0
    private var listW = 0
    private var listH = 0
    private var cmdFieldX = 0
    private var cmdFieldW = 0
    private var removeBtnX = 0

    override fun init() {
        for (e in CommandKeys.all()) {
            keys.add(e.key())
            commands.add(e.command())
        }

        listH = ROW_H * MAX_VISIBLE
        panelH = 50 + listH + 46
        panelX = (this.width - panelW) / 2
        panelY = max(8, (this.height - panelH) / 2)

        listX = panelX + 14
        listY = panelY + 44
        listW = panelW - 28
        cmdFieldX = listX + KEY_BTN_W + 6
        cmdFieldW = listW - KEY_BTN_W - 6 - REMOVE_BTN_W - 6
        removeBtnX = listX + listW - REMOVE_BTN_W

        rebuildRows()
    }

    private fun persist() {
        val list = ArrayList<CommandKeys.Entry>()
        for (i in keys.indices) list.add(CommandKeys.Entry(keys[i], commands[i]))
        CommandKeys.replaceAll(list)
    }

    private fun rebuildRows() {
        clearChildren()

        for (i in keys.indices) {
            val rowTop = listY + i * ROW_H - scroll
            if (rowTop + ROW_H < listY || rowTop > listY + listH) continue
            val idx = i

            val k = keys[i]
            val label = if (capturingIndex != null && capturingIndex == i) {
                "> Press a key <"
            } else if (k == InputUtil.UNKNOWN_KEY) {
                "Unbound"
            } else {
                k.localizedText.string
            }
            addDrawableChild(
                ButtonWidget.builder(Text.literal(label)) {
                    capturingIndex = idx
                    rebuildRows()
                }
                    .dimensions(listX, rowTop + 3, KEY_BTN_W, 18).build()
            )

            val cmdField = TextFieldWidget(this.textRenderer, cmdFieldX, rowTop + 3, cmdFieldW, 18, Text.literal("Command"))
            cmdField.setMaxLength(256)
            cmdField.text = commands[i]
            cmdField.setChangedListener { s ->
                commands[idx] = s
                persist()
            }
            addDrawableChild(cmdField)

            addDrawableChild(
                ButtonWidget.builder(Text.literal("§cX")) {
                    keys.removeAt(idx)
                    commands.removeAt(idx)
                    persist()
                    rebuildRows()
                }
                    .dimensions(removeBtnX, rowTop + 3, REMOVE_BTN_W, 18).build()
            )
        }

        val btnY = listY + listH + 8
        addDrawableChild(
            ButtonWidget.builder(Text.literal("+ Add Command Key")) {
                keys.add(InputUtil.UNKNOWN_KEY)
                commands.add("")
                persist()
                rebuildRows()
            }
                .dimensions(panelX + 14, btnY, 160, 20).build()
        )
        addDrawableChild(
            ButtonWidget.builder(Text.literal("Done")) { close() }
                .dimensions(panelX + panelW - 14 - 70, btnY, 70, 20).build()
        )
    }

    override fun renderBackground(ctx: DrawContext, mouseX: Int, mouseY: Int, delta: Float) {
        super.renderBackground(ctx, mouseX, mouseY, delta)
        ctx.fill(panelX - 1, panelY - 1, panelX + panelW + 1, panelY + panelH + 1, BORDER)
        ctx.fill(panelX, panelY, panelX + panelW, panelY + panelH, BG_PANEL)
        ctx.fill(panelX, panelY, panelX + panelW, panelY + 22, BG_SECTION)
        ctx.fill(panelX, panelY + 22, panelX + panelW, panelY + 23, ACCENT)
        ctx.drawCenteredTextWithShadow(this.textRenderer, "§b§lCommand Keys", panelX + panelW / 2, panelY + 7, 0xFFFFFF)
        ctx.drawTextWithShadow(
            this.textRenderer,
            "§7Click a key box, then press a key or click a mouse button §8(Esc to unbind)",
            panelX + 14, panelY + 30, TEXT_HINT
        )
        ctx.fill(listX - 2, listY - 2, listX + listW + 2, listY + listH + 2, LIST_BG)
    }

    override fun mouseClicked(click: Click, bl: Boolean): Boolean {
        val idx = capturingIndex
        if (idx != null) {
            keys[idx] = InputUtil.Type.MOUSE.createFromCode(click.button())
            capturingIndex = null
            persist()
            rebuildRows()
            return true
        }
        return super.mouseClicked(click, bl)
    }

    override fun keyPressed(input: KeyInput): Boolean {
        val idx = capturingIndex
        if (idx != null) {
            keys[idx] = if (input.key() == GLFW.GLFW_KEY_ESCAPE) InputUtil.UNKNOWN_KEY else InputUtil.fromKeyCode(input)
            capturingIndex = null
            persist()
            rebuildRows()
            return true
        }
        return super.keyPressed(input)
    }

    override fun mouseScrolled(mouseX: Double, mouseY: Double, horizontalAmount: Double, verticalAmount: Double): Boolean {
        val maxScroll = max(0, keys.size * ROW_H - listH)
        scroll = max(0, min(maxScroll, scroll - (verticalAmount * ROW_H).toInt()))
        rebuildRows()
        return true
    }
}
